server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -2277,14 +2295,21 @@ public ClusterState execute(ClusterState currentState) {
                 }
                 // Snapshot ids that will have to be physically deleted from the repository
                 final Set<SnapshotId> snapshotIdsRequiringCleanup = new HashSet<>(snapshotIds);
+
+                final List<Runnable> completeAbortedAssignedQueuedRunnables = new ArrayList<>();
                 final SnapshotsInProgress updatedSnapshots = snapshotsInProgress.createCopyWithUpdatedEntriesForRepo(
                     projectId,
                     repositoryName,
                     snapshotsInProgress.forRepo(projectId, repositoryName).stream().map(existing -> {
                         if (existing.state() == SnapshotsInProgress.State.STARTED
                             && snapshotIdsRequiringCleanup.contains(existing.snapshot().getSnapshotId())) {
                             // snapshot is started - mark every non completed shard as aborted
-                            final SnapshotsInProgress.Entry abortedEntry = existing.abort();
+                            final SnapshotsInProgress.Entry abortedEntry = existing.abort(
+                                currentState.nodes().getLocalNodeId(),
+                                ((shardId, shardSnapshotStatus) -> completeAbortedAssignedQueuedRunnables.add(
+                                    () -> innerUpdateSnapshotState(existing.snapshot(), shardId, shardSnapshotStatus)
```

Comment:
Apply updates as ShardSnapshotUpdate tasks. I think this is a better approach than trying to update them locally which would require duplicate logics from the update task executors. We already have some duplication in processExternalChanges and there is a [TODO](https://github.com/elastic/elasticsearch/blob/80faf64ceedab7f2e360e752cd3624f1eab3c0f6/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L1060-L1062) suggesting refactoring it away.